File: Android/app/src/main/java/com/auralis/crisisconnect/ai/CrisisSentinelOnlineClient.kt
package com.auralis.crisisconnect.ai

import com.auralis.crisisconnect.BuildConfig
import com.auralis.crisisconnect.network.PinnedOkHttpClient
import com.google.android.gms.tasks.Task
import java.io.IOException
import java.util.concurrent.TimeUnit
import com.google.firebase.auth.FirebaseAuth
import kotlinx.coroutines.suspendCancellableCoroutine
import okhttp3.Call
import okhttp3.Callback
import okhttp3.HttpUrl
import okhttp3.HttpUrl.Companion.toHttpUrlOrNull
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import okhttp3.Response
import org.json.JSONArray
import org.json.JSONObject
import kotlin.coroutines.resume
import kotlin.coroutines.resumeWithException

/** One chat turn in the dashboard chat API's wire format. */
data class CrisisSentinelOnlineMessage(
    val sender: String,
    val text: String
) {
    companion object {
        const val SENDER_USER = "user"
        const val SENDER_BOT = "bot"
    }
}

class CrisisSentinelOnlineException(
    val kind: Kind,
    val httpStatus: Int? = null,
    message: String
) : Exception(message) {
    enum class Kind {
        /** No signed-in (non-anonymous) Firebase user, or the server rejected the token. */
        Unauthorized,

        /** The account lacks dashboard access (role/panel) server-side. */
        Forbidden,

        /** The agency panel's AI quota is exhausted. */
        Quota,

        /** Transport-level failure (unreachable, reset, timeout between events). */
        Network,

        /** The server answered but with an unexpected status or payload. */
        Server,

        /** The panel has no AI provider configured (server ended with needsApiKey). */
        NotConfigured
    }
}

/**
 * Client for the web dashboard's Crisis Sentinel chat API: `POST /api/chat/stream` (SSE) and
 * `GET /api/chat/providers`. Field-team only: callers must gate on the role certificate before
 * invoking; the server enforces the same gate via the Firestore user document (role + agency
 * panel).
 *
 * The base URL is the dashboard's Cloud Run origin, not a certificate-pinned custom domain:
 * `*.run.app` leaves are Google-managed and rotate like `*.googleapis.com`, so the NSC
 * "custom hosts must be pinned" rule does not apply (same brick-risk rationale documented in
 * network_security_config.xml).
 */
class CrisisSentinelOnlineClient(
    baseUrl: String = BuildConfig.CRISIS_SENTINEL_ONLINE_BASE_URL
) {
    private val baseHttpUrl = baseUrl.trim().trimEnd('/').toHttpUrlOrNull()
        ?.takeIf { it.isHttps }

    private val client: OkHttpClient = PinnedOkHttpClient.newClient().newBuilder()
        .connectTimeout(15, TimeUnit.SECONDS)
        // Watchdog between SSE frames; the server emits an early `status` event, so a healthy
        // stream never stays silent this long. The overall cap is the ViewModel's timeout.
        .readTimeout(75, TimeUnit.SECONDS)
        .build()

    /**
     * Sends the conversation and streams the assistant reply. [onDelta] receives the full
     * accumulated text after each server delta (invoked from an OkHttp thread). [modelChoice]
     * pins a provider/model; null lets the server pick its default. Cancelling the calling
     * coroutine aborts the HTTP call immediately.
     */
    suspend fun streamChat(
        messages: List<CrisisSentinelOnlineMessage>,
        modelChoice: CrisisSentinelOnlineModelChoice? = null,
        onDelta: (accumulated: String) -> Unit
    ): CrisisSentinelOnlineStreamResult = withIdToken { token ->
        executeStream(token, messages, modelChoice, onDelta)
    }

    /** Providers/models the signed-in user's panel can use. Server omits unconfigured ones. */
    suspend fun fetchProviders(): List<CrisisSentinelOnlineProvider> = withIdToken { token ->
        executeFetchProviders(token)
    }

    /**
     * Runs [block] with a Firebase ID token; on an Unauthorized failure retries exactly once
     * with a force-refreshed token (covers the cached-token revocation window). Throws
     * [CrisisSentinelOnlineException] with kind Unauthorized when no non-anonymous user exists.
     */
    private suspend fun <T> withIdToken(block: suspend (token: String) -> T): T {
        val user = FirebaseAuth.getInstance().currentUser?.takeUnless { it.isAnonymous }
            ?: throw CrisisSentinelOnlineException(
                kind = CrisisSentinelOnlineException.Kind.Unauthorized,
                message = "No signed-in Firebase user"
            )
        val token = fetchIdToken(user::getIdToken, forceRefresh = false)
        return try {
            block(token)
        } catch (unauthorized: CrisisSentinelOnlineException) {
            if (unauthorized.kind != CrisisSentinelOnlineException.Kind.Unauthorized) throw unauthorized
            val freshToken = fetchIdToken(user::getIdToken, forceRefresh = true)
            block(freshToken)
        }
    }

    private suspend fun fetchIdToken(
        getIdToken: (Boolean) -> Task<com.google.firebase.auth.GetTokenResult>,
        forceRefresh: Boolean
    ): String {
        val token = runCatching { getIdToken(forceRefresh).awaitTask().token?.trim() }
            .getOrNull()
        if (token.isNullOrBlank()) {
            throw CrisisSentinelOnlineException(
                kind = CrisisSentinelOnlineException.Kind.Unauthorized,
                message = "Could not obtain Firebase ID token"
            )
        }
        return token
    }

    private fun endpoint(vararg segments: String): HttpUrl {
        val base = baseHttpUrl ?: throw CrisisSentinelOnlineException(
            kind = CrisisSentinelOnlineException.Kind.NotConfigured,
            message = "Online chat base URL is not configured"
        )
        return base.newBuilder().apply { segments.forEach(::addPathSegments) }.build()
    }

    private suspend fun executeFetchProviders(token: String): List<CrisisSentinelOnlineProvider> {
        val request = Request.Builder()
            .url(endpoint("api/chat/providers"))
            .header("Authorization", "Bearer $token")
            .get()
            .build()
        val body = executeCall(request) { response ->
            if (!response.isSuccessful) {
                val detail = runCatching {
                    JSONObject(response.body?.string().orEmpty()).optString("error")
                }.getOrNull().orEmpty()
                throw mapStatus(response.code, detail.ifBlank { "HTTP ${response.code}" })
            }
            response.body?.string().orEmpty()
        }
        return parseProviders(body)
    }

    private suspend fun executeStream(
        token: String,
        messages: List<CrisisSentinelOnlineMessage>,
        modelChoice: CrisisSentinelOnlineModelChoice?,
        onDelta: (accumulated: String) -> Unit
    ): CrisisSentinelOnlineStreamResult {
        val messagesJson = JSONArray()
        messages.takeLast(MAX_MESSAGES).forEach { message ->
            messagesJson.put(
                JSONObject()
                    .put("sender", message.sender)
                    .put("text", message.text.take(MAX_MESSAGE_CHARS))
            )
        }
        val body = JSONObject()
            .put("messages", messagesJson)
            .put("context", JSONObject())
        if (modelChoice != null) {
            body.put("providerId", modelChoice.providerId)
            body.put("model", modelChoice.modelId)
        }
        val request = Request.Builder()
            .url(endpoint("api/chat/stream"))
            .header("Authorization", "Bearer $token")
            .header("Accept", "text/event-stream")
            .post(body.toString().toRequestBody(jsonMediaType))
            .build()

        return executeCall(request) { response -> readStream(response, onDelta) }
    }

    /** Enqueues [request] and runs [handler] on OkHttp's thread with coroutine cancellation. */
    private suspend fun <T> executeCall(
        request: Request,
        handler: (Response) -> T
    ): T = suspendCancellableCoroutine { continuation ->
        val call = client.newCall(request)
        continuation.invokeOnCancellation { call.cancel() }
        call.enqueue(object : Callback {
            override fun onFailure(call: Call, e: IOException) {
                if (!continuation.isActive) return
                continuation.resumeWithException(
                    CrisisSentinelOnlineException(
                        kind = CrisisSentinelOnlineException.Kind.Network,
                        message = e.message ?: "Network failure"
                    )
                )
            }

            override fun onResponse(call: Call, response: Response) {
                val outcome = runCatching { response.use(handler) }
                if (!continuation.isActive) return
                outcome
                    .onSuccess(continuation::resume)
                    .onFailure { throwable ->
                        continuation.resumeWithException(
                            throwable as? CrisisSentinelOnlineException
                                ?: CrisisSentinelOnlineException(
                                    kind = CrisisSentinelOnlineException.Kind.Network,
                                    message = throwable.message ?: "Request failed"
                                )
                        )
                    }
            }
        })
    }

    private fun readStream(
        response: Response,
        onDelta: (String) -> Unit
    ): CrisisSentinelOnlineStreamResult {
        if (!response.isSuccessful) {
            val detail = runCatching {
                JSONObject(response.body?.string().orEmpty()).optString("error")
            }.getOrNull().orEmpty()
            throw mapStatus(response.code, detail.ifBlank { "HTTP ${response.code}" })
        }
        val source = response.body?.source()
            ?: throw CrisisSentinelOnlineException(
                kind = CrisisSentinelOnlineException.Kind.Server,
                message = "Empty response body"
            )

        val accumulated = StringBuilder()
        var providerIdFromMetadata: String? = null
        var eventName = ""
        val dataLines = mutableListOf<String>()

        while (true) {
            val line = source.readUtf8Line() ?: break
            when {
                line.isEmpty() -> {
                    val data = dataLines.joinToString("\n")
                    dataLines.clear()
                    val name = eventName
                    eventName = ""
                    when (name) {
                        "metadata" -> {
                            providerIdFromMetadata = runCatching {
                                JSONObject(data).optJSONObject("_provider")?.optString("id")
                            }.getOrNull()?.takeIf { it.isNotBlank() }
                        }

                        "text_delta" -> {
                            val delta = runCatching { JSONObject(data).optString("delta") }
                                .getOrDefault("")
                            if (delta.isNotEmpty()) {
                                accumulated.append(delta)
                                onDelta(accumulated.toString())
                            }
                        }

                        "done" -> return finishFromDone(
                            data = data,
                            streamedText = accumulated.toString().trim(),
                            providerIdFromMetadata = providerIdFromMetadata
                        )

                        "error" -> {
                            val json = runCatching { JSONObject(data) }.getOrNull()
                            val status = json?.optInt("status", 0)?.takeIf { it > 0 }
                            throw mapStatus(
                                status ?: 0,
                                json?.optString("error").orEmpty().ifBlank { "Stream error" }
                            )
                        }
                        // "status" phases and "tool_call"/"tool_result" notifications are
                        // informational; the substance arrives as text and the done payload.
                    }
                }

                line.startsWith("event:") -> eventName = line.removePrefix("event:").trim()
                line.startsWith("data:") -> dataLines.add(line.removePrefix("data:").trim())
            }
        }

        // EOF without a `done` event: keep whatever streamed rather than discarding a
        // mostly-complete reply over a dropped connection tail.
        val text = accumulated.toString().trim()
        if (text.isNotEmpty()) return CrisisSentinelOnlineStreamResult(text = text)
        throw CrisisSentinelOnlineException(
            kind = CrisisSentinelOnlineException.Kind.Network,
            message = "Stream ended before any reply"
        )
    }

    private fun finishFromDone(
        data: String,
        streamedText: String,
        providerIdFromMetadata: String?
    ): CrisisSentinelOnlineStreamResult {
        val payload = runCatching { JSONObject(data).optJSONObject("data") }.getOrNull()
        if (payload?.optBoolean("needsApiKey") == true) {
            throw CrisisSentinelOnlineException(
                kind = CrisisSentinelOnlineException.Kind.NotConfigured,
                message = "Panel has no AI provider configured"
            )
        }

        val candidate = payload?.optJSONArray("candidates")?.optJSONObject(0)
        val parts = candidate?.optJSONObject("content")?.optJSONArray("parts")

        val partText = StringBuilder()
        val mapPoints = mutableListOf<CrisisSentinelMapPoint>()
        val unsupportedTools = mutableListOf<String>()
        for (index in 0 until (parts?.length() ?: 0)) {
            val part = parts?.optJSONObject(index) ?: continue
            part.optString("text").takeIf { it.isNotBlank() }?.let(partText::append)
            val functionCall = part.optJSONObject("functionCall") ?: continue
            val toolName = functionCall.optString("name")
            if (toolName == TOOL_SHOW_LOCATION_ON_MAP) {
                mapPoints += parseMapPoints(functionCall.optJSONObject("args"))
            } else if (toolName.isNotBlank()) {
                unsupportedTools += toolName
            }
        }

        val text = streamedText.ifEmpty { partText.toString().trim() }
        val cardJson = payload?.optJSONObject("_card")?.toString()
        if (text.isEmpty() && cardJson == null && mapPoints.isEmpty() && unsupportedTools.isEmpty()) {
            throw CrisisSentinelOnlineException(
                kind = CrisisSentinelOnlineException.Kind.Server,
                message = "Stream finished without reply content"
            )
        }
        return CrisisSentinelOnlineStreamResult(
            text = text,
            modelName = payload?.optString("model")?.takeIf { it.isNotBlank() }
                ?: providerIdFromMetadata,
            cardJson = cardJson,
            mapPoints = mapPoints,
            unsupportedTools = unsupportedTools,
            groundingMetadataJson = candidate?.optJSONObject("groundingMetadata")?.toString()
        )
    }

    private fun parseMapPoints(args: JSONObject?): List<CrisisSentinelMapPoint> {
        val points = args?.optJSONArray("points") ?: return emptyList()
        return buildList {
            for (index in 0 until points.length()) {
                val point = points.optJSONObject(index) ?: continue
                val lat = point.optDouble("lat", Double.NaN)
                val lng = point.optDouble("lng", Double.NaN)
                if (lat.isNaN() || lng.isNaN()) continue
                add(
                    CrisisSentinelMapPoint(
                        lat = lat,
                        lng = lng,
                        label = point.optString("label"),
                        details = point.optString("details").takeIf { it.isNotBlank() },
                        type = point.optString("type").takeIf { it.isNotBlank() }
                    )
                )
            }
        }
    }

    private fun parseProviders(body: String): List<CrisisSentinelOnlineProvider> {
        val providers = runCatching { JSONObject(body).optJSONArray("providers") }.getOrNull()
            ?: throw CrisisSentinelOnlineException(
                kind = CrisisSentinelOnlineException.Kind.Server,
                message = "Malformed providers response"
            )
        return buildList {
            for (index in 0 until providers.length()) {
                val provider = providers.optJSONObject(index) ?: continue
                val id = provider.optString("id").takeIf { it.isNotBlank() } ?: continue
                val models = buildList {
                    val modelsJson = provider.optJSONArray("models")
                    for (modelIndex in 0 until (modelsJson?.length() ?: 0)) {
                        val model = modelsJson?.optJSONObject(modelIndex) ?: continue
                        val modelId = model.optString("id").takeIf { it.isNotBlank() } ?: continue
                        add(
                            CrisisSentinelOnlineModelInfo(
                                id = modelId,
                                label = model.optString("label").ifBlank { modelId },
                                contextWindow = model.optLong("contextWindow", -1L)
                                    .takeIf { it > 0L }
                            )
                        )
                    }
                }
                add(
                    CrisisSentinelOnlineProvider(
                        id = id,
                        label = provider.optString("label").ifBlank { id },
                        defaultModel = provider.optString("defaultModel")
                            .takeIf { it.isNotBlank() },
                        source = provider.optString("source"),
                        models = models
                    )
                )
            }
        }
    }

    private fun mapStatus(status: Int, message: String): CrisisSentinelOnlineException {
        val kind = when (status) {
            401 -> CrisisSentinelOnlineException.Kind.Unauthorized
            403 -> CrisisSentinelOnlineException.Kind.Forbidden
            429 -> CrisisSentinelOnlineException.Kind.Quota
            else -> CrisisSentinelOnlineException.Kind.Server
        }
        return CrisisSentinelOnlineException(kind = kind, httpStatus = status, message = message)
    }

    private companion object {
        // Client-side truncation keeps requests inside the server's validation limits
        // (50 messages / 20k chars per message) with margin.
        const val MAX_MESSAGES = 30
        const val MAX_MESSAGE_CHARS = 20_000
        const val TOOL_SHOW_LOCATION_ON_MAP = "showLocationOnMap"
        val jsonMediaType = "application/json; charset=utf-8".toMediaType()
    }
}

private suspend fun <T> Task<T>.awaitTask(): T = suspendCancellableCoroutine { continuation ->
    addOnCompleteListener { task ->
        if (task.isSuccessful) {
            val result = task.result
            if (result != null) {
                continuation.resume(result)
            } else {
                continuation.resumeWithException(IllegalStateException("Task returned null"))
            }
        } else {
            continuation.resumeWithException(
                task.exception ?: IllegalStateException("Task failed without exception")
            )
        }
    }
}

File: iOS/Crisis Connect/Features/CrisisSentinel/CrisisSentinelOnlineClient.swift
//
//  CrisisSentinelOnlineClient.swift
//  Crisis Connect
//
//  Client for the web dashboard's Crisis Sentinel chat API — iOS port of Android's
//  `CrisisSentinelOnlineClient`: `POST /api/chat/stream` (SSE) and `GET /api/chat/providers`.
//  Field-team/admin only: callers gate on the rescue-certificate role before invoking; the server
//  enforces the same gate from the Firestore user document (role + agency panel) and applies the
//  panel's AI quota.
//
//  The base URL is the dashboard's Cloud Run origin, NOT the hosting domain — Firebase Hosting
//  buffers SSE, so the stream must hit Cloud Run directly (same origin the web client hard-codes).
//

import Foundation
import FirebaseAuth

struct CrisisSentinelOnlineMessage {
    static let senderUser = "user"
    static let senderBot = "bot"
    let sender: String
    let text: String
}

struct CrisisSentinelMapPoint: Equatable, Codable {
    let lat: Double
    let lng: Double
    let label: String
    var details: String? = nil
    var type: String? = nil
}

struct CrisisSentinelOnlineModelInfo: Identifiable, Equatable {
    let id: String
    let label: String
    var contextWindow: Int64? = nil
}

struct CrisisSentinelOnlineProvider: Identifiable, Equatable {
    let id: String
    let label: String
    var defaultModel: String? = nil
    var source: String = ""
    var models: [CrisisSentinelOnlineModelInfo] = []
}

struct CrisisSentinelOnlineModelChoice: Equatable, Codable {
    let providerId: String
    let modelId: String
}

struct CrisisSentinelOnlineStreamResult {
    let text: String
    var modelName: String? = nil
    var cardJson: String? = nil
    var mapPoints: [CrisisSentinelMapPoint] = []
    var unsupportedTools: [String] = []
    var groundingMetadataJson: String? = nil
}

struct CrisisSentinelOnlineError: Error {
    enum Kind {
        /// No signed-in (non-anonymous) Firebase user, or the server rejected the token.
        case unauthorized
        /// The account lacks dashboard access (role/panel) server-side.
        case forbidden
        /// The agency panel's AI quota is exhausted.
        case quota
        /// Transport-level failure (unreachable, reset, timeout between events).
        case network
        /// The server answered but with an unexpected status or payload.
        case server
        /// The panel has no AI provider configured (server ended with needsApiKey).
        case notConfigured
    }

    let kind: Kind
    var httpStatus: Int? = nil
    let message: String
}

final class CrisisSentinelOnlineClient {
    // Same margins as Android: keep requests inside the server's validation limits
    // (50 messages / 20k chars per message).
    private static let maxMessages = 30
    private static let maxMessageChars = 20_000
    private static let toolShowLocationOnMap = "showLocationOnMap"

    /// The dashboard's Cloud Run origin. Override per-build via Info.plist key
    /// `CrisisSentinelOnlineBaseURL` when staging; the default is production.
    static var baseURL: URL? = {
        if let raw = Bundle.main.object(forInfoDictionaryKey: "CrisisSentinelOnlineBaseURL") as? String,
           let url = URL(string: raw.trimmingCharacters(in: .whitespacesAndNewlines)),
           url.scheme == "https" {
            return url
        }
        return URL(string: "https://ssrcrisisconnect1-jxxgznalnq-uc.a.run.app")
    }()

    private let session: URLSession = {
        let config = URLSessionConfiguration.default
        config.timeoutIntervalForRequest = 75 // watchdog between SSE frames (server pings early)
        config.timeoutIntervalForResource = 300
        return URLSession(configuration: config)
    }()

    /// Sends the conversation and streams the assistant reply. `onDelta` receives the FULL
    /// accumulated text after each server delta (hop to the main actor before touching UI).
    /// `modelChoice` pins a provider/model; nil lets the server pick its default.
    func streamChat(
        messages: [CrisisSentinelOnlineMessage],
        modelChoice: CrisisSentinelOnlineModelChoice? = nil,
        onDelta: @escaping (String) -> Void
    ) async throws -> CrisisSentinelOnlineStreamResult {
        try await withIdToken { token in
            try await self.executeStream(token: token, messages: messages, modelChoice: modelChoice, onDelta: onDelta)
        }
    }

    /// Providers/models the signed-in user's panel can use. Server omits unconfigured ones.
    func fetchProviders() async throws -> [CrisisSentinelOnlineProvider] {
        try await withIdToken { token in
            try await self.executeFetchProviders(token: token)
        }
    }

    // MARK: - Auth

    /// Runs `block` with a Firebase ID token; on an Unauthorized failure retries exactly once with
    /// a force-refreshed token (covers the cached-token revocation window).
    private func withIdToken<T>(_ block: (String) async throws -> T) async throws -> T {
        guard let user = Auth.auth().currentUser, !user.isAnonymous else {
            throw CrisisSentinelOnlineError(kind: .unauthorized, message: "No signed-in Firebase user")
        }
        let token = try await fetchIdToken(user: user, forceRefresh: false)
        do {
            return try await block(token)
        } catch let error as CrisisSentinelOnlineError where error.kind == .unauthorized {
            let fresh = try await fetchIdToken(user: user, forceRefresh: true)
            return try await block(fresh)
        }
    }

    private func fetchIdToken(user: User, forceRefresh: Bool) async throws -> String {
        let token = try? await user.getIDTokenResult(forcingRefresh: forceRefresh).token
        guard let token = token?.trimmingCharacters(in: .whitespacesAndNewlines), !token.isEmpty else {
            throw CrisisSentinelOnlineError(kind: .unauthorized, message: "Could not obtain Firebase ID token")
        }
        return token
    }

    private func endpoint(_ path: String) throws -> URL {
        guard let base = Self.baseURL else {
            throw CrisisSentinelOnlineError(kind: .notConfigured, message: "Online chat base URL is not configured")
        }
        return base.appendingPathComponent(path)
    }

    // MARK: - Providers

    private func executeFetchProviders(token: String) async throws -> [CrisisSentinelOnlineProvider] {
        var request = URLRequest(url: try endpoint("api/chat/providers"))
        request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")

        let (data, response): (Data, URLResponse)
        do {
            (data, response) = try await session.data(for: request)
        } catch {
            throw CrisisSentinelOnlineError(kind: .network, message: error.localizedDescription)
        }
        let status = (response as? HTTPURLResponse)?.statusCode ?? 0
        guard (200..<300).contains(status) else {
            throw Self.mapStatus(status, detail: Self.errorDetail(from: data))
        }
        return Self.parseProviders(data)
    }

    // MARK: - Stream

    private func executeStream(
        token: String,
        messages: [CrisisSentinelOnlineMessage],
        modelChoice: CrisisSentinelOnlineModelChoice?,
        onDelta: @escaping (String) -> Void
    ) async throws -> CrisisSentinelOnlineStreamResult {
        var body: [String: Any] = [
            "messages": messages.suffix(Self.maxMessages).map {
                ["sender": $0.sender, "text": String($0.text.prefix(Self.maxMessageChars))]
            },
            "context": [String: Any](),
        ]
        if let modelChoice {
            body["providerId"] = modelChoice.providerId
            body["model"] = modelChoice.modelId
        }

        var request = URLRequest(url: try endpoint("api/chat/stream"))
        request.httpMethod = "POST"
        request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
        request.setValue("text/event-stream", forHTTPHeaderField: "Accept")
        request.setValue("application/json; charset=utf-8", forHTTPHeaderField: "Content-Type")
        request.httpBody = try? JSONSerialization.data(withJSONObject: body)

        let (bytes, response): (URLSession.AsyncBytes, URLResponse)
        do {
            (bytes, response) = try await session.bytes(for: request)
        } catch {
            throw CrisisSentinelOnlineError(kind: .network, message: error.localizedDescription)
        }
        let status = (response as? HTTPURLResponse)?.statusCode ?? 0
        guard (200..<300).contains(status) else {
            var errorBody = Data()
            for try await byte in bytes.prefix(64 * 1024) { errorBody.append(byte) }
            throw Self.mapStatus(status, detail: Self.errorDetail(from: errorBody))
        }

        // SSE state machine (pure, so it's unit-testable) — identical event handling to Android.
        var parser = SSEEventParser()
        do {
            for try await line in bytes.lines {
                switch try parser.consume(line: line) {
                case .none:
                    break
                case .delta(let accumulated):
                    onDelta(accumulated)
                case .done(let result):
                    return result
                }
            }
        } catch let error as CrisisSentinelOnlineError {
            throw error
        } catch {
            // Mid-stream transport failure: keep a mostly-complete reply rather than discarding it.
            if let partial = parser.partialResult() { return partial }
            throw CrisisSentinelOnlineError(kind: .network, message: error.localizedDescription)
        }
        // EOF without a `done` event.
        if let partial = parser.partialResult() { return partial }
        throw CrisisSentinelOnlineError(kind: .network, message: "Stream ended before any reply")
    }

    // MARK: - Payload parsing (byte-parity with Android)

    static func finishFromDone(
        data: String,
        streamedText: String,
        providerIdFromMetadata: String?
    ) throws -> CrisisSentinelOnlineStreamResult {
        let payload = json(data)?["data"] as? [String: Any]
        if payload?["needsApiKey"] as? Bool == true {
            throw CrisisSentinelOnlineError(kind: .notConfigured, message: "Panel has no AI provider configured")
        }

        let candidate = (payload?["candidates"] as? [[String: Any]])?.first
        let parts = (candidate?["content"] as? [String: Any])?["parts"] as? [[String: Any]] ?? []

        var partText = ""
        var mapPoints: [CrisisSentinelMapPoint] = []
        var unsupportedTools: [String] = []
        for part in parts {
            if let text = part["text"] as? String, !text.trimmingCharacters(in: .whitespaces).isEmpty {
                partText += text
            }
            guard let functionCall = part["functionCall"] as? [String: Any] else { continue }
            let toolName = functionCall["name"] as? String ?? ""
            if toolName == toolShowLocationOnMap {
                mapPoints += parseMapPoints(functionCall["args"] as? [String: Any])
            } else if !toolName.isEmpty {
                unsupportedTools.append(toolName)
            }
        }

        let text = streamedText.isEmpty
            ? partText.trimmingCharacters(in: .whitespacesAndNewlines)
            : streamedText
        var cardJson: String?
        if let card = payload?["_card"] as? [String: Any],
           let cardData = try? JSONSerialization.data(withJSONObject: card) {
            cardJson = String(data: cardData, encoding: .utf8)
        }
        if text.isEmpty && cardJson == nil && mapPoints.isEmpty && unsupportedTools.isEmpty {
            throw CrisisSentinelOnlineError(kind: .server, message: "Stream finished without reply content")
        }

        var groundingJson: String?
        if let grounding = candidate?["groundingMetadata"] as? [String: Any],
           let groundingData = try? JSONSerialization.data(withJSONObject: grounding) {
            groundingJson = String(data: groundingData, encoding: .utf8)
        }
        let model = (payload?["model"] as? String).flatMap { $0.isEmpty ? nil : $0 }
        return CrisisSentinelOnlineStreamResult(
            text: text,
            modelName: model ?? providerIdFromMetadata,
            cardJson: cardJson,
            mapPoints: mapPoints,
            unsupportedTools: unsupportedTools,
            groundingMetadataJson: groundingJson
        )
    }

    private static func parseMapPoints(_ args: [String: Any]?) -> [CrisisSentinelMapPoint] {
        guard let points = args?["points"] as? [[String: Any]] else { return [] }
        return points.compactMap { point in
            guard let lat = (point["lat"] as? NSNumber)?.doubleValue,
                  let lng = (point["lng"] as? NSNumber)?.doubleValue else { return nil }
            return CrisisSentinelMapPoint(
                lat: lat,
                lng: lng,
                label: point["label"] as? String ?? "",
                details: (point["details"] as? String).flatMap { $0.isEmpty ? nil : $0 },
                type: (point["type"] as? String).flatMap { $0.isEmpty ? nil : $0 }
            )
        }
    }

    static func parseProviders(_ data: Data) -> [CrisisSentinelOnlineProvider] {
        guard let root = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
              let providers = root["providers"] as? [[String: Any]] else {
            return []
        }
        return providers.compactMap { provider in
            guard let id = (provider["id"] as? String), !id.isEmpty else { return nil }
            let models = (provider["models"] as? [[String: Any]] ?? []).compactMap { model -> CrisisSentinelOnlineModelInfo? in
                guard let modelId = model["id"] as? String, !modelId.isEmpty else { return nil }
                let label = (model["label"] as? String).flatMap { $0.isEmpty ? nil : $0 } ?? modelId
                let context = (model["contextWindow"] as? NSNumber)?.int64Value
                return CrisisSentinelOnlineModelInfo(
                    id: modelId, label: label, contextWindow: (context ?? 0) > 0 ? context : nil
                )
            }
            return CrisisSentinelOnlineProvider(
                id: id,
                label: (provider["label"] as? String).flatMap { $0.isEmpty ? nil : $0 } ?? id,
                defaultModel: (provider["defaultModel"] as? String).flatMap { $0.isEmpty ? nil : $0 },
                source: provider["source"] as? String ?? "",
                models: models
            )
        }
    }

    static func json(_ raw: String) -> [String: Any]? {
        guard let data = raw.data(using: .utf8) else { return nil }
        return (try? JSONSerialization.jsonObject(with: data)) as? [String: Any]
    }

    private static func errorDetail(from data: Data) -> String {
        let detail = ((try? JSONSerialization.jsonObject(with: data)) as? [String: Any])?["error"] as? String
        return detail?.isEmpty == false ? detail! : ""
    }

    static func mapStatus(_ status: Int, detail: String) -> CrisisSentinelOnlineError {
        let kind: CrisisSentinelOnlineError.Kind
        switch status {
        case 401: kind = .unauthorized
        case 403: kind = .forbidden
        case 429: kind = .quota
        default: kind = .server
        }
        return CrisisSentinelOnlineError(
            kind: kind,
            httpStatus: status,
            message: detail.isEmpty ? "HTTP \(status)" : detail
        )
    }
}

/// Pure SSE event state machine — feed it lines (from bytes.lines or a test), and it accumulates
/// text deltas, resolves the metadata provider id, and finalizes on the `done` event (or throws on
/// `error`). Extracted from the stream loop so the parsing is unit-testable without a network.
struct SSEEventParser {
    enum Outcome {
        case none
        /// The full accumulated reply text after a delta.
        case delta(String)
        case done(CrisisSentinelOnlineStreamResult)
    }

    private var accumulated = ""
    private var providerIdFromMetadata: String?
    private var eventName = ""
    private var dataLines: [String] = []

    mutating func consume(line: String) throws -> Outcome {
        if line.isEmpty {
            let data = dataLines.joined(separator: "\n")
            dataLines.removeAll()
            let name = eventName
            eventName = ""
            switch name {
            case "metadata":
                if let json = CrisisSentinelOnlineClient.json(data),
                   let provider = json["_provider"] as? [String: Any],
                   let id = (provider["id"] as? String)?.trimmingCharacters(in: .whitespaces),
                   !id.isEmpty {
                    providerIdFromMetadata = id
                }
                return .none
            case "text_delta":
                if let delta = CrisisSentinelOnlineClient.json(data)?["delta"] as? String, !delta.isEmpty {
                    accumulated += delta
                    return .delta(accumulated)
                }
                return .none
            case "done":
                return .done(try CrisisSentinelOnlineClient.finishFromDone(
                    data: data,
                    streamedText: accumulated.trimmingCharacters(in: .whitespacesAndNewlines),
                    providerIdFromMetadata: providerIdFromMetadata
                ))
            case "error":
                let json = CrisisSentinelOnlineClient.json(data)
                let status = (json?["status"] as? NSNumber)?.intValue ?? 0
                let message = (json?["error"] as? String).flatMap { $0.isEmpty ? nil : $0 } ?? "Stream error"
                throw CrisisSentinelOnlineClient.mapStatus(status, detail: message)
            default:
                return .none // "status" / "tool_call" / "tool_result" are informational.
            }
        } else if line.hasPrefix("event:") {
            eventName = String(line.dropFirst("event:".count)).trimmingCharacters(in: .whitespaces)
        } else if line.hasPrefix("data:") {
            dataLines.append(String(line.dropFirst("data:".count)).trimmingCharacters(in: .whitespaces))
        }
        return .none
    }

    /// On EOF or a mid-stream transport failure, keep a mostly-complete reply rather than
    /// discarding it; nil when nothing streamed.
    func partialResult() -> CrisisSentinelOnlineStreamResult? {
        let text = accumulated.trimmingCharacters(in: .whitespacesAndNewlines)
        return text.isEmpty ? nil : CrisisSentinelOnlineStreamResult(text: text)
    }
}
